### PR TITLE
Attach to pid fixes with threading. Fixes #1542,#1520

### LINKED ISSUE #1542: If threading is not imported in main thread, attach to pid does not work properly.
I was doing some manual testing for the attach to pid and in this situation, if the user never imported `threading`, we don't currently set the tracing properly.

The main reason for this is that `threading` seems to expect that it's **imported** from the actual main thread and not from another thread (because presumably one would import `threading` to create a thread in the first place) and it creates the initial `_MainThread` with the current thread (even if that's not the actual main thread).

Now, in the attach to pid, what happens is that we create a thread on the target process and then call the `settrace` in that thread... and if `threading` was never imported it'll consider that dummy thread as the main thread because we import `threading` in the dummy thread and the actual main thread will not be visible to `threading.enumerate()`.

I still don't know the best way to overcome this...

Additional notes:
- This happens in Python 2 and Python 3.
- We can detect that if `threading` not in `sys.modules` during the attach.
- In Linux we don't create a thread during the attach and use the current running thread, so, this shouldn't be a problem (although it could be an issue if the user has multiple threads not started through `threading` and we happen to stop in the wrong thread, then we'd import `threading` in the wrong thread).

### PR DESCRIPTION
**Pull request by fabioz**
As a note, Linux (on Azure) is failing because there's no gdb and I'm not exactly sure how to do that on Azure (see: https://github.com/microsoft/ptvsd/pull/1582/files#diff-354f30a63fb0907d4ad57269548329e3R12 for the changes needed to run on travis -- i.e.: install gdb and set the `ptrace_scope` to `0`).

I'm currently taking a look at the failures on Python 2.7 (windows and mac), so, this still can't be merged.

### CODE REVIEW

**Comment by karthiknadig** [comment]
Updated the build to have gdb.

**Comment by karthiknadig** [comment]
With the latest run. Linux passes, windows 2.7 is the flask issue.

**Comment by karthiknadig** [comment]
Just merged the temp fix for the flask issue.

File: src/ptvsd/_vendored/pydevd/pydevd_attach_to_process/windows/attach.cpp [hunk 1]
```diff
@@ -350,50 +100,34 @@ std::wstring GetCurrentModuleFilename() {
...
-HANDLE g_initedEvent;
-int AttachCallback(void *initThreads) {
+
+int AttachCallback(void *voidInitializeThreadingInfo) {
     // initialize us for threading, this will acquire the GIL if not already created, and is a nop if the GIL is created.
     // This leaves us in the proper state when we return back to the runtime whether the GIL was created or not before
     // we were called.
-    ((PyEval_Lock*)initThreads)();
-    SetEvent(g_initedEvent);
+    InitializeThreadingInfo* initializeThreadingInfo = (InitializeThreadingInfo*)voidInitializeThreadingInfo;
```
Comment (karthiknadig) [inline src/ptvsd/_vendored/pydevd/pydevd_attach_to_process/windows/attach.cpp]: NIT: reinterpret_cast<InitializeThreadingInfo*>()

File: src/ptvsd/_vendored/pydevd/pydevd_attach_to_process/windows/attach.cpp [hunk 2]
```diff
@@ -495,7 +229,7 @@ void SuspendThreads(ThreadMap &suspendedThreads, Py_AddPendingCall* addPendingCa
                                 GetThreadContext(hThread, &context);
 
 #if defined(_X86_)
-                                if(context.Eip >= *((DWORD*)addPendingCall) && context.Eip <= (*((DWORD*)addPendingCall)) + 0x100) {
+                                if (context.Eip >= *((DWORD*)addPendingCall) && context.Eip <= (*((DWORD*)addPendingCall)) + 0x100) {
```
Comment (karthiknadig) [inline src/ptvsd/_vendored/pydevd/pydevd_attach_to_process/windows/attach.cpp]: reinterpret_cast

**Comment by fabioz** [comment]
> Other than `tests/func/test_attach.py::test_attaching_by_pid_no_threading[code-attach_pid]` test failure in Windows 2.7. rest looks good.

I just did another change to improve that (so, if it passes I'll merge to master).

**Comment by fabioz** [comment]
All green 🎉

**Pull request merged**